emu/src/render/ppu.rs

```diff
@@ -61,42 +68,74 @@ impl PixelProcessUnit {
                 todo!("BG_MODE 2 not implemented yet")
             }
             3 => {
-                // Bitmap mode
+                acquire_lock!(self.internal_memory, memory => {
+                    // Bitmap mode
+                    for y in 0..LCD_HEIGHT {
+                        for x in 0..LCD_WIDTH {
+                            let index_color = (y * LCD_WIDTH + x) * 2;
+
+                            let color_array = [
+                                memory.video_ram[index_color],
+                                memory.video_ram[index_color + 1],
+                            ];
+
+                            let color: Color = [color_array[0], color_array[1]].into();
+
+                            acquire_lock!(self.gba_lcd, gba_lcd => { gba_lcd.set_pixel(x, y, color); })
+                        }
+                    }
+                });
+            }
+            4 => {
+                // 06000000-06009FFF for Frame 0
+                // 0600A000-06013FFF for Frame 1
+                let selected_frame = acquire_lock!(self.internal_memory, memory => {
+                    memory.lcd_registers.get_frame_select()
+                });
+
                 for y in 0..LCD_HEIGHT {
                     for x in 0..LCD_WIDTH {
-                        let index_color = (y * LCD_WIDTH + x) * 2;
-                        let color: Color = [
-                            memory.video_ram[index_color],
-                            memory.video_ram[index_color + 1],
-                        ]
-                        .into();
-
-                        gba_lcd.set_pixel(x, y, color);
+                        let index_mem =
+                            (y * LCD_WIDTH + x) + (selected_frame * LCD_HEIGHT * LCD_WIDTH);
+
+                        let index_palette = acquire_lock!(self.internal_memory, memory => {
+                            memory.video_ram[index_mem]
+                        });
+
+                        let color: Color = self
+                            .get_color_from_full_palette(index_palette.into(), &PaletteType::BG);
+
+                        acquire_lock!(self.gba_lcd, gba_lcd => { gba_lcd.set_pixel(x, y, color); })
```

Comment:
Can we lock outside of the for so that we don't lock every iteration? I think it's also semantically more correct because we lock the access to memory/lcd while we refresh it and then we release it

Reply:
The problem is the method in line 106.

Reply:
Mhh I get it but I don't know, I would try to rethink the why it works. I know we should not focus on performance right now but I think it's better to address this now that we're very early stage

Reply:
I'll think to a solution so.

Reply:
This macro only add a strict scope of the locks? Because maybe we can achieve this in other way if it is the only limit of our design now